tls/s2n_cipher_suites.c

```diff
@@ -1168,13 +1216,179 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
+            if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+                return false;
+            }
+        }
+
+        /**
+        *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+        *# The server MUST ensure that it selects a compatible PSK
+        *# (if any) and cipher suite.
+        **/
+        if (conn->psk_params.chosen_psk != NULL) {
+            if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+                return false;
+            }
+        }
+
+        return true;
+}
+
+/**
+ * @brief Same as s2n_wire_ciphers_contain() except that it sets the match's index in the client wire to the index var.
+ * 
+ * @param match Cipher suite IANA which we wish to find a match for
+ * @param wire Client wire
+ * @param count Number of entries
+ * @param cipher_suite_len Length offset
+ * @return int Index of the match in the client wire. If a match was not found then -1 is returned.
+ */
+static int s2n_wire_ciphers_has_server_cipher_at(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+{
+    int negative_one = -1;
+    for (uint32_t i = 0; i < count; i++) {
+        const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
+
+        if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return i;
+        }
+    }
+
+    return negative_one;
+}
+
+/**
+ * @brief This is the logic for parsing a list of s2n_cipher_suites* and selecting a cipher_suite accounting
+ *        for equal preference grouping. This method assumes that the server side cipher suite preference is
+ *        correctly formatted eg for every group_start delimiter, there exists a following group_end delimiter.
+ *        TODO: Add reference to documentation and add basic validation
+ * 
+ * @param conn s2n_connection is used to ensure the negotiated cipher is valid for the connection
+ * @param wire client wire containing cipher suite IANAs
+ * @param count Number of entries
+ * @param cipher_suite_len Cipher suite length offset
+ * @return int Server's index of the negotiated cipher suite. If no cipher suite was selected then -1 is returned.
+ */
+static int s2n_get_negotiated_server_index(struct s2n_connection* conn,
+                                           const uint8_t* wire,
+                                           uint32_t count,
+                                           uint32_t cipher_suite_len) {
+    POSIX_ENSURE_REF(wire);
+    POSIX_ENSURE_REF(conn);
+
+    const struct s2n_security_policy *security_policy;
+    POSIX_GUARD(s2n_connection_get_security_policy(conn, &security_policy));
+    const struct s2n_cipher_preferences* cipher_preferences = security_policy->cipher_preferences;
+    POSIX_ENSURE_REF(cipher_preferences);
+
+    bool in_group = false;
+    int negotiated_client_index = count;
+    int negotiated_server_index = -1, negotiated_server_highest_vers_match_index = -1;
```

Comment:
I'm not really following what 'negotiated_server_highest_vers_match_index' means

Reply:
It's the same as the old `higher_vers_match` var. The only difference now is that it is index to the cipher.

Reply:
Yes, they are indices as indicated by the suffix. I can just remove the prefix so that it becomes `highest_vers_match_index`. That way the name is pretty much identical to the original variables except their indices instead of pointers to cipher suites